controllers/bucket.go

```diff
@@ -0,0 +1,259 @@
+}
+
+var (
+	// ErrorOutOfRange is error when value is out of range
+	ErrorOutOfRange error = errors.New("Value out of range")
+)
+
+// GetBuckets factory to construct buckets from provided scaling intervals
+func GetBuckets(hvpa *hvpav1alpha1.Hvpa, currentReplicas int32) (bucketList map[string]EffectiveScalingIntervals, currentBucket *EffectiveScalingInterval, err error) {
+	var (
+		containerMapBucketMap                map[string][]*EffectiveScalingInterval = make(map[string][]*EffectiveScalingInterval)
+		containerBucketMap                   map[string]EffectiveScalingIntervals   = make(map[string]EffectiveScalingIntervals)
+		currBucket                           *EffectiveScalingInterval
+		minReplicasGlobal                    int32 = 1
+		firstContainerName                   string
+		firstContainerFirstIntervalBucketLen int
+	)
+	scalingOverlap := hvpa.Spec.ScalingIntervalsOverlap
+
+	if hvpa.Spec.Hpa.Template.Spec.MinReplicas != nil {
+		minReplicasGlobal = *hvpa.Spec.Hpa.Template.Spec.MinReplicas
+	}
+
+	// Since minAllowed is taken as the min resource value for the first interval for each container,
+	// the buckets generated using first scaleInterval can be different for each container.
+	// But rest of the buckets are going to be the same for all the containers because for rest of the scaleIntervals,
+	// the min resource values are calculated using max resource of each interval and scalingOverlap which is common for all containers.
+	for i := range hvpa.Spec.Vpa.Template.Spec.ResourcePolicy.ContainerPolicies {
+		container := &hvpa.Spec.Vpa.Template.Spec.ResourcePolicy.ContainerPolicies[i]
+		if container.Mode != nil && *container.Mode == vpa_api.ContainerScalingModeOff {
+			continue
+		}
+		minCPU, minMemory := int64(0), int64(0)
+		containerMapBucketMap[container.ContainerName] = make([]*EffectiveScalingInterval, 0)
+		if i == 0 {
+			firstContainerName = container.ContainerName
+		}
+		minReplicas := minReplicasGlobal
+
+		if container.MinAllowed != nil {
+			minCPU = container.MinAllowed.Cpu().MilliValue()
+			minMemory = container.MinAllowed.Memory().MilliValue()
+		}
+		for j, scaleInterval := range hvpa.Spec.ScaleIntervals {
+			bucket, curr, err := getLinearBuckets(container.MaxAllowed, scaleInterval, scalingOverlap, &minCPU, &minMemory, &minReplicas, currentReplicas, hvpa.Namespace+"/"+hvpa.Name)
+			if err != nil {
+				return nil, nil, err
+			}
+			if i == 0 && j == 0 {
+				firstContainerFirstIntervalBucketLen = len(bucket)
+			}
+			containerMapBucketMap[container.ContainerName] = append(containerMapBucketMap[container.ContainerName], bucket...)
+			if curr != nil {
+				currBucket = curr
+			}
+			if i != 0 {
+				// For 2nd container onwards, since rest of the buckets are same as for first container, append and break
+				containerMapBucketMap[container.ContainerName] = append(containerMapBucketMap[container.ContainerName], containerMapBucketMap[firstContainerName][firstContainerFirstIntervalBucketLen:]...)
```

Comment:
As mentioned in the start of the `for` loop:
>// Since minAllowed is taken as the min resource value for the first interval for each container,
	// the buckets generated using first scaleInterval can be different for each container.
	// But rest of the buckets are going to be the same for all the containers because for rest of the scaleIntervals,
	// the min resource values are calculated using max resource of each interval and scalingOverlap which is common for all containers.

Reply:
Thanks! Sorry, I missed that.

Hmm... this seems to be a hole in the [design section](https://github.com/gardener/hvpa-controller/blob/master/docs/proposals/HVPA-EP-0001/HVPA-EP-001.md#design) of the proposal. Do you think it makes sense to allow `ScalingInterval.MaxResources` to optionally specify percentage (of VPA `maxAllowed` or the absolute `maxAllowed` of the last interval) in addition to absolute values? I am fine, picking this later too. Just thinking loudly.

Reply:
>Do you think it makes sense to allow ScalingInterval.MaxResources to optionally specify percentage

This would result in `scalingIntervals` for all containers in pod to be "proportionally similar". So a container with range of 1G to 20G memory, and another with range of 100M to 2G memory will have similar weightage when it comes to determining final replica to scale to or when it comes to "choosing a bucket/`scaleInterval`".

As per current proposal, **if `scaleIntervals` are provided keeping the scaling needs of larger container**, the smaller container's scaling will be inhibited, because its recommendation may not cross larger bucket boundaries that easily.
So larger container should end up dominating number of replicas, and smaller container would be only scaling vertically according to those number of replicas and its VPA recommendation.

Even though it is not symmetric/consistent from POV of all containers, IMHO, it should work out better for scaling scenarios than having "proportionally similar" `scalingIntervals`. WDYT?

This assumes that the resource usage of sidecar never exceeds the usage of main container.

Reply:
> Even though it is not symmetric/consistent from POV of all containers, IMHO, it should work out better for scaling scenarios than having "proportionally similar" `scalingIntervals`. WDYT?

👍  I was only thinking aloud. Makes sense to be to stick to the current proposal and its design and only introduce more complexity only if there are genuine use-cases.

PS: I loved the term _proportionally similar_ 😀